mixing/mix.py

```diff
@@ -0,0 +1,285 @@
+import logging, shutil
+from typing import List
+from tqdm import tqdm
+import gc
+
+import torch
+from transformers import (
+    set_seed,
+    AutoConfig,
+    AutoModelForCausalLM,
+    LlamaForCausalLM,
+    MistralForCausalLM,
+    MixtralForCausalLM,
+    MixtralConfig
+)
+from datasets import load_from_disk
+
+from mixing.models.mixllama import MixLlamaConfig, MixLlamaForCausalLM
+
+MODELS_MOE_CLASS = {
+    LlamaForCausalLM: (MixLlamaForCausalLM, MixLlamaConfig),
+    MistralForCausalLM: (MixtralForCausalLM, MixtralConfig),
+}
+
+ROUTERS_NAME = {
+    MixLlamaForCausalLM: ".router.",
+    MixtralForCausalLM: ".gate."
+}
+
+EXPERTS_KEYS_MAPPING = {
+    MixLlamaForCausalLM: {
+        "mlp": "mlp.experts",
+        "q_proj": "q_proj.experts",
+        "k_proj": "k_proj.experts",
+        "v_proj": "v_proj.experts",
+    },
+    MixtralForCausalLM: {
+        "mlp": "block_sparse_moe.experts",
+    }
+}
+
+# dont know why mixtral change naming
+KEYS_REMAP = {
+    MixtralForCausalLM: {
+        "mlp":{
+            "w1":"gate_proj",
+            "w3":"up_proj",
+            "w2":"down_proj"
+        }
+    }
+}
+
+@torch.no_grad()
+def mix(
+    base_model:str,
+    ingredients:List[str],
+    modules_to_mix:List[str],
+    positive_tokens:List[str]=[],
+    num_samples:int=1000,
+    num_experts_per_tok:int=2,
+):
+    
+    set_seed(1399)
+
+    logging.info("loading base model...")
+    config_base = AutoConfig.from_pretrained(base_model)
+    model_base = AutoModelForCausalLM.from_pretrained(base_model)
+    model_base_type = type(model_base)
+    sd_base = model_base.state_dict()
+    MOE_MODEL_CLS, MOE_CFG_CLS = MODELS_MOE_CLASS[model_base_type]
```

Comment:
local variables in python should be lowercase

Reply:
done

Reply:
I still see a lot of upper-case local variables throughout the code

Reply:
last commit is not reflected in this comment section